libp2p/NodeTable.cpp

```diff
@@ -102,34 +115,28 @@ bool NodeTable::addNode(Node const& _node, NodeRelation _relation)
         return true;
     }
 
-    if (!_node.endpoint || !_node.id)
-    {
-        LOG(m_logger) << "Supplied node has an invalid endpoint (" << _node.endpoint << ") or id ("
-                      << _node.id << "). Skipping adding to node table.";
-        return false;
-    }
-
+    bool bFound = false;
+    shared_ptr<NodeEntry> newNodeEntry = nullptr;
     DEV_GUARDED(x_nodes)
     {
         auto const it = m_allNodes.find(_node.id);
         if (it != m_allNodes.end())
-        {
-            LOG(m_logger) << "Node " << _node.id << "@" << _node.endpoint
-                          << " is already in the node table";
-            return true;
-        }
+            bFound = true;
+        else
+            m_allNodes[_node.id] = make_shared<NodeEntry>(m_hostNodeID, _node.id, _node.endpoint);
+        newNodeEntry = m_allNodes[_node.id];
     }
 
-    if (m_hostNodeID == _node.id)
-    {
-        LOG(m_logger) << "Skip adding self to node table (" << _node.id << ")";
-        return false;
-    }
+    // Log here to avoid holding the x_nodes mutex longer than necessary
+    if (bFound)
+        LOG(m_logger) << "Node " << _node.id << "@" << _node.endpoint
+                      << " is already in the node table";
+    else
+        LOG(m_logger) << "Pending node " << _node.id << "@" << _node.endpoint;
+
+    if (!newNodeEntry->hasValidEndpointProof())
```

Comment:
Maybe we also could simplify this function by removing `if (_relation == Known)` part, then adding ` && _relation != Known` on this line.
Then we'd also need to add
```
  if (_relation == Known)
     newNodeEntry->lastPongReceivedTime = RLPXDatagramFace::secondsSinceEpoch();
```

Reply:
Previously the case for `Known` was completely disjoint so I planned to move it to a separate function. Now it has some sanity check which are good to have so it is not so simple any more.